pilot/pkg/model/context.go

```diff
@@ -66,10 +66,36 @@ type Proxy struct {
 	// "default.svc.cluster.local")
 	Domain string
 
+	// Version represents the binary version of the proxy
+	// e.g., 0.8.0
+	Version ProxyVersion
+
 	// Metadata key-value pairs extending the Node identifier
 	Metadata map[string]string
 }
 
+// ProxyVersion indicates the version of the data plane proxy we are talking to.
```

Comment:
As I mentioned - I'm not sure what is the right way to handle this, the proxy version doesn't really help since it's not something pilot can really process.  We don't support a single sidecar that has linear versions - the sidecar can be a custom build with extra modules, from a remote repo using its own versioning.

How about not designing a new API and model in 1.0 branch - and just use the presence of the env variable in 1.0 to differentiate 0.8.0 ( which will never have this set ) ?

Reply:
We can still add capabilities by using x.y.z-tags. This is not exposing any api. It’s simply parsing the metadata field into some struct that we can use internally. Plugins are welcome to add additional info and parse it in any way they want. For networking code in open source, we simply need to track the istio versions so that we can emit the right type of Envoy configs for things like filter chain matching on src/dst ip ranges etc. a true version info would be one where Envoy sends list of filters with versions and capabilities it has. This is a simpler version of that. We could do a lot of things with just this. 

If we implement unimplemented pieces of Envoy proto, we need to know of the Envoy will respect it or ignore it, in the main networking code. Such things cannot be converted into tags. There would be hundreds of such tags then (eg filterchainipsupport,rediscds,mongofault,etc) one for each thing we add to Envoy. The numbering scheme is a way to mark these features on a linear semver scale, sort of like kernel code. 

I don’t want to have dummy env vars that will get removed in next cycle. People could still use new istioctl with 0.8 proxy, or stock Envoy. Using a value in the environment variable transfers the burden to the operator making custom changes to clearly communicate the major istio release with which the proxy is expected to work.